### ISSUE TITLE
Feature request: Dropping image onto the ui creates markdown cell to display the image

### ISSUE THREAD

**Issue by fiedl** [nteract/archived-desktop-app#5081]
I would like to be able to drag-and-drop an image file (png/jpg/...) into the nteract ui to display the image in the notebook. This could be used, for example, for sketches, schematics, or other resources.

![drop-image-onto-nteract](https://user-images.githubusercontent.com/1679688/80892354-aacb8080-8cc9-11ea-87e3-d85533ae971b.gif)

Possible approach: Listen to the drop event, and create a new markdown cell before the cell the image has been dropped upon. The new markdown cell get the content

    ![](/path/to/the/image)

where `/path/to/the/image` is the path to the image from the drop event.

I would be willing to invest some time to implement this, but I would need some pointers:

- Where would I add the event listener?

  ```js
  document.addEventListener('drop', function(event) {});
  ```

- What is the api to add a markdown cell with a certain content?

- How could I add the new markdown cell *before* the cell the drop has occurred on?

Any help would be very much appreciated. Thanks!

**Comment by captainsafia** [comment]
Thanks for taking the time to open this issue and for offering to work on the implementation, @fiedl!

This idea is similar to the earlier proposed feature request to support pasting images into Markdown cells, https://github.com/nteract/nteract/issues/4963. We might be able to tackle both issues at once.

> Where would I add the event listener?

Since this handler is specific to the desktop app and registered globally, it should be added [here](https://github.com/nteract/nteract/blob/master/applications/desktop/src/notebook/global-events.ts#L43).

> What is the api to add a markdown cell with a certain content?

nteract represents the open notebook document inside a Redux store. To modify the notebook document, there are API actions that can be invoked. The particular ones that will be of interest to you are the `createCellBelow`/`createCellAbove` actions which can be invoked like so.

```js
import { actions } from "@nteract/actions";

actions.createCellBelow({
  cellType: "markdown",
  contentRef: "notebookContentRef",
  source: "Some initial source"
});
```

The most helpful way to learn about these actions is to go through the source code and look for examples of their invocation.

> How could I add the new markdown cell before the cell the drop has occurred on?

Good question! This is where things start to get a little bit interesting.

Currently, nteract maintains a unique identifier that refers to the currently focused cell in the Redux store. In fact, the `createCellBelow` function will add the new cell below the currently focused cell by default. So, it's relatively straightforward to add the new cell before the currently focused one.

Determining which cell is closest to the drop area is going to be a little bit more interesting. I don't have any good (performant and maintainable) ideas for how to do this at the moment. 

Let me know if this information is helpful. Feel free to ask more questions on the Slack (linked in the README).

**Comment by jarmokivekas** [comment]
:+1: because I would love to have this features. I always end up with "TODO: add image" everywhere because I'm too lazy to copy file paths :smile:

**Comment by fiedl** [comment]
Thank you, @captainsafia, for your detailed help!

I've implemented a first attempt, but I would have a few questions, if I may:

- `actions.createCellBelow` works fine, but `actions.createCellAbove`, which I wanted to use, does ignore the `source` argument. For the moment, I've just used `createCellBelow` instead. But is there a reason for [`createCellAbove`](https://github.com/nteract/nteract/blob/e756d849974103b3efd2519e14a1c2d457d12c2e/packages/reducers/src/core/entities/contents/notebook.ts#L517) not to consider the `source` parameter?
- I've added a small add-on: When holding down the <kbd>option</kbd> (<kbd>alt</kbd>) key while dropping the images, I'm copying the images into the notebook directory. In macOS, holding the 
<kbd>option</kbd> key while hovering actually shows a little plus icon indicating a "copy" action. But I'm unsure about linux and windows. I think, holding down the <kbd>ctrl</kbd> key might be more common there. Is there a more abstract way to implement this switch using the operating systems' defaults?

  ![Bildschirmfoto 2020-05-11 um 21 36 39](https://user-images.githubusercontent.com/1679688/81603990-83626b00-93cf-11ea-8257-2699ec5fcebe.png)

- When copying images to the notebook directory, I'm silently not overwriting existing files. Is there a common way to show some kind of warning banner "file … already exists"?

- I wanted to extract the functionality into a separate file `insert-images.ts` and was wondering where to put it. There is an [`applications/desktop/src/notebook/epics`](https://github.com/nteract/nteract/tree/master/applications/desktop/src/notebook/epics) directory, which also caught my attention. I've never used redux epics before. Would there be any benefit of extracting this functionality out into an epic? Otherwise, I'd just extract the code into a regular module in `applications/desktop/src/notebook/insert-images.ts`.

- I've run the app using `yarn dev`:

  ```bash
  [2020-05-11 21:13:23] fiedl@fiedl-mbp ~/code/nteract/applications/desktop fiedl/drop-images ⚡
  ▶ yarn dev
  ```
  
  Is there a way to run this with `--watch` in order to pick up code changes without re-running `yarn dev`?

Thanks!

**Comment by captainsafia** [comment]
> But is there a reason for createCellAbove not to consider the source parameter?

There isn't. In fact, after another conversation I had with a dev I'm thinking that the `createCellBelow` and `createCellAbove` actions should be set up to take an `ImmutableCell` object. I can submit a PR for this.

> Is there a more abstract way to implement this switch using the operating systems' defaults?

Good question! Depending on what the keyboard shortcut does you might be able to use Electron's accelerators ([ref](https://www.electronjs.org/docs/api/accelerator)).

> When copying images to the notebook directory, I'm silently not overwriting existing files. Is there a common way to show some kind of warning banner "file … already exists"?

Yep, you can take a look at our notifications API. You can look at invocations of the `sendNotification` function for a reference on how this works ([ref](https://github.com/nteract/nteract/search?l=TypeScript&q=sendNotification)).

> Would there be any benefit of extracting this functionality out into an epic? Otherwise, I'd just extract the code into a regular module in applications/desktop/src/notebook/insert-images.ts.

It depends on the particular contents of the implementation. If it is written as a Redux epic, you can move it to that directory. If not, it's fine to move it to the path that you specified above.

> Is there a way to run this with --watch in order to pick up code changes without re-running yarn dev?

Yes, you can try following the steps outlined in [this doc](https://github.com/nteract/nteract/blob/master/CONTRIBUTING.md#validating-in-the-nteract-desktop-app). Let me know if this is not working for you.

**Comment by fiedl** [comment]
> Is there a more abstract way to implement this switch using the operating systems' defaults?

It turns out, the native [`event.dataTransfer.effectAllowed`](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/Drag_operations#drageffects) already provides a nice interface for that.

I've used this mechanism to give the user access to the different insertion methods:

- native "drag to copy" operation: copy the image to the notebook directory and insert as markdown image
 

  ![](https://user-images.githubusercontent.com/1679688/81603990-83626b00-93cf-11ea-8257-2699ec5fcebe.png)


- native "drag to link" operation: don't copy the image, but use the original image path to insert as markdown image

  ![Bildschirmfoto 2020-05-17 um 17 13 25](https://user-images.githubusercontent.com/1679688/82152477-ba1e0280-9861-11ea-8e34-a13b2d0660dc.png)

- native "drag" operation without modifiers: embed the image as base64 source directly into the markdown cell as suggested by https://github.com/nteract/nteract/issues/4963#issuecomment-628275602.

**Comment by fiedl** [comment]
Thank you, again, @captainsafia, for your help!

> When copying images to the notebook directory, I'm silently not overwriting existing files. Is there a common way to show some kind of warning banner "file … already exists"?

Thanks for suggesting `sendNotification`. This works fine!

![Bildschirmfoto 2020-05-17 um 16 15 03](https://user-images.githubusercontent.com/1679688/82150829-948cfb00-9859-11ea-9821-8bde7605e08a.png)

I've also implemented a "Replace" button as notification action.

However, after replacing the file, the markdown cell stills shows the old image. Do you know a way to reload the cell in order to display the new image? Unfortunately, even removing and re-creating the markdown cell by hand, does not reload the new image. Maybe there is some kind of caching mechanism involved?

**Comment by captainsafia** [comment]
Wow! Great work polishing up all these small details. I'm excited to see the final pull request.

> Do you know a way to reload the cell in order to display the new image?

Does the image get updated if you click on the Markdown cell, make another edit, then toggle to preview mode by clicking outside the well? If so, that might indicate that the issue is in the re-render logic in React. Because the images both have the same filename, the string content of the markdown cell doesn't change when the image is replaced on disk. As a result, React doesn't re-render the component.

> Unfortunately, even removing and re-creating the markdown cell by hand, does not reload the new image.

The plot thickens. If this is the case, then it likely has nothing to do with React's rendering at all. I'm not aware of any caching that Electron includes by default. What do you see in the "network" tab under developer tools when you trigger the replace or re-create the cell-like you did above? There's a chance that Electron might not re-request the image once it has been loaded.

**Comment by fiedl** [comment]
> Does the image get updated if you click on the Markdown cell, make another edit, then toggle to preview mode by clicking outside the well?

No, it doesn't. The network tab does not show a new activity in this case.

Also when copying the content to a new markdown cell, the new cell does not show the new image and does not create network activity.

Restarting `yarn spawn` and re-opening the saved notebook causes the new image to be displayed.

> There's a chance that Electron might not re-request the image once it has been loaded.

Looks like it.

**Comment by captainsafia** [comment]
> Looks like it.

In that case, we might want to try using [this approach](https://stackoverflow.com/a/1077051) to trigger a re-fetch of the image.
